pkg/agent/controller/egress/egress_controller.go

```diff
@@ -905,3 +905,39 @@ func (c *EgressController) deleteEgressGroup(group *cpv1b2.EgressGroup) {
 	delete(c.egressGroups, group.Name)
 	c.queue.Add(group.Name)
 }
+
+// GetEgressIPFromMark returns the Egress IP associated with the snatMark.
+func (c *EgressController) GetEgressIPFromMark(snatMark uint32) (string, error) {
+	c.egressIPStatesMutex.Lock()
+	defer c.egressIPStatesMutex.Unlock()
+	for _, e := range c.egressIPStates {
+		if e.mark == snatMark {
+			return e.egressIP.String(), nil
+		}
+	}
+	return "", fmt.Errorf("no EgressIP associated with mark %v", snatMark)
+}
+
+// GetEgressPolicy returns effective Egress applied on a Pod.
+func (c *EgressController) GetEgressPolicy(ns, podName string) (string, error) {
+	c.egressBindingsMutex.RLock()
+	defer c.egressBindingsMutex.RUnlock()
+	pod := k8s.NamespacedName(ns, podName)
+	binding, exists := c.egressBindings[pod]
+	if !exists {
+		return "", fmt.Errorf("no Egress binded with %v", pod)
+	}
+	return binding.effectiveEgress, nil
+}
+
+// EgressPacket returns true if packet is for Egress.
+func (c *EgressController) EgressPacket(tunnelDstIP string) bool {
+	c.egressIPStatesMutex.Lock()
+	defer c.egressIPStatesMutex.Unlock()
+	for egressIP := range c.egressIPStates {
+		if egressIP == tunnelDstIP {
```

Comment:
checking whether a key exists in a map should be O(1), but was implemented as O(N) here.